### PR TITLE
Remove rest of PostgresDriver functionality

### PR DESCRIPTION
**Pull request by philandstuff**
This moves the RecordIndex stuff out from PostgresDriver{NonTransactional,Transactional} into new classes that implement the new RecordIndex interface, just like #127 (for entries) and #128 (for items).

This marks the third of a three-part refactoring, and is where we get the benefits of deleting `PostgresDriver*`, `BackingStoreDriver` and so on. 🙌

### CODE REVIEW

**Comment by michaelabenyohai** [comment]
[country-er.txt](https://github.com/openregister/openregister-java/files/649541/country-er.txt) - this file successfully loads using `curl -X POST --data-binary @country-er.rsf http://localhost:8080/load-rsf --header "Content-Type:application/uk-gov-rsf"`

[country-er-mid-assert.txt](https://github.com/openregister/openregister-java/files/649540/country-er-mid-assert.txt) - this file does not successfully load using `curl -X POST --data-binary @country-er-mid-assert.rsf http://localhost:8080/load-rsf --header "Content-Type:application/uk-gov-rsf"`

The only difference is that the second file has an extra `assert-root-hash` in the middle of the file. It may have something to with that we now use the same `VerifiableLog` object for each `assert-root-hash` command whereas we used to `new` a `VerifiableLog` for each `assert-root-hash` command.

I also don't believe we have a test for this case and we should (sorry)!

The error I get is:

```
ERROR [2016-12-13 17:09:39,600] uk.gov.register.providers.ThrowableExceptionMapper: Uncaught exception: {}
! java.lang.IllegalStateException: iterator is closed
! at org.skife.jdbi.v2.ResultSetResultIterator.next(ResultSetResultIterator.java:86)
! at org.skife.jdbi.v2.sqlobject.ResultReturnThing$IteratorResultReturnThing$1.next(ResultReturnThing.java:201)
! at uk.gov.register.db.EntryIteratorDAO.findByEntryNumber(EntryIteratorDAO.java:26)
! at uk.gov.register.db.EntryMerkleLeafStore.getLeafValue(EntryMerkleLeafStore.java:22)
! at uk.gov.verifiablelog.VerifiableLog.computeSubtreeHash(VerifiableLog.java:128)
! at uk.gov.verifiablelog.VerifiableLog.subtreeHash(VerifiableLog.java:144)
! at uk.gov.verifiablelog.VerifiableLog.computeSubtreeHash(VerifiableLog.java:131)
! at uk.gov.verifiablelog.VerifiableLog.subtreeHash(VerifiableLog.java:144)
! at uk.gov.verifiablelog.VerifiableLog.computeSubtreeHash(VerifiableLog.java:131)
! at uk.gov.verifiablelog.VerifiableLog.subtreeHash(VerifiableLog.java:144)
! at uk.gov.verifiablelog.VerifiableLog.computeSubtreeHash(VerifiableLog.java:131)
! at uk.gov.verifiablelog.VerifiableLog.subtreeHash(VerifiableLog.java:144)
! at uk.gov.verifiablelog.VerifiableLog.computeSubtreeHash(VerifiableLog.java:131)
! at uk.gov.verifiablelog.VerifiableLog.subtreeHash(VerifiableLog.java:144)
! at uk.gov.verifiablelog.VerifiableLog.computeSubtreeHash(VerifiableLog.java:131)
! at uk.gov.verifiablelog.VerifiableLog.subtreeHash(VerifiableLog.java:144)
! at uk.gov.verifiablelog.VerifiableLog.computeSubtreeHash(VerifiableLog.java:131)
! at uk.gov.verifiablelog.VerifiableLog.subtreeHash(VerifiableLog.java:144)
! at uk.gov.verifiablelog.VerifiableLog.computeSubtreeHash(VerifiableLog.java:132)
! at uk.gov.verifiablelog.VerifiableLog.subtreeHash(VerifiableLog.java:144)
! at uk.gov.verifiablelog.VerifiableLog.computeSubtreeHash(VerifiableLog.java:131)
! at uk.gov.verifiablelog.VerifiableLog.subtreeHash(VerifiableLog.java:144)
! at uk.gov.verifiablelog.VerifiableLog.getCurrentRootHash(VerifiableLog.java:48)
! at uk.gov.register.db.TransactionalEntryLog.getRegisterProof(TransactionalEntryLog.java:96)
! at uk.gov.register.core.PostgresRegister.getRegisterProof(PostgresRegister.java:132)
! at uk.gov.register.serialization.AssertRootHashCommand.execute(AssertRootHashCommand.java:19)
! at uk.gov.register.service.RegisterSerialisationFormatService.lambda$mintRegisterComponents$1(RegisterSerialisationFormatService.java:74)
! ... 84 common frames omitted
```

**Comment by michaelabenyohai** [comment]
My last comment should be addressed by https://github.com/openregister/openregister-java/pull/133. Once that is merged (after I write some tests) we should be able to rebase this and no longer see the exception.

File: src/main/java/uk/gov/register/service/RegisterService.java [hunk 1]
```diff
@@ -37,16 +45,26 @@ public RegisterService(RegisterData registerData, DBI dbi, MemoizationStore memo
...
+        useTransaction(dbi, handle -> {
 
             String registerName = registerData.getRegister().getRegisterName();
-            Handle handle = postgresDriver.getHandle();
-            TransactionalEntryLog entryLog = new TransactionalEntryLog(transactionalMemoizationStore, handle.attach(EntryQueryDAO.class), handle.attach(EntryDAO.class));
-            TransactionalItemStore itemStore = new TransactionalItemStore(handle.attach(ItemDAO.class), handle.attach(ItemQueryDAO.class), itemValidator);
-            Register register = new PostgresRegister(() -> registerName, postgresDriver, registerFieldsConfiguration, entryLog, itemStore);
+            EntryLog entryLog = new TransactionalEntryLog(transactionalMemoizationStore, handle.attach(EntryQueryDAO.class), handle.attach(EntryDAO.class));
+            ItemStore itemStore = new TransactionalItemStore(handle.attach(ItemDAO.class), handle.attach(ItemQueryDAO.class), itemValidator);
+            RecordIndex onDemandRecordIndex = new TransactionalRecordIndex(handle.attach(RecordQueryDAO.class), handle.attach(CurrentKeysUpdateDAO.class));
```
Comment (philandstuff) [inline src/main/java/uk/gov/register/service/RegisterService.java]: Yup, will do

File: src/test/java/uk/gov/register/store/postgres/PostgresDriverTransactionalTest.java [hunk 2]
```diff
@@ -1,89 +0,0 @@
...
-        PostgresDriverTransactional postgresDriver = new PostgresDriverTransactional(
-                handle, h -> recordQueryDAO, h -> currentKeysUpdateDAO);
-
-        postgresDriver.insertRecord("DE", 1);
-
-        assertThat(currentKeys, is(empty()));
-    }
-
-    @Test
-    public void getRecordShouldAlwaysCommitStagedData() {
```
Comment (philandstuff) [inline src/test/java/uk/gov/register/store/postgres/PostgresDriverTransactionalTest.java]: I have written tests that assert the correct output is produced but are deliberately agnostic to whether the data is flushed to the database or not.

As long as the overall transaction succeeds (tested in PostgresRegisterTransactionalFunctionalTest) that should be all we need.

**Pull request merged**